packages/gatsby-source-wordpress/src/store.ts

```diff
@@ -1,14 +1,46 @@
-import { init } from "@rematch/core"
+import { RematchStore, init } from "@rematch/core"
 import immerPlugin from "@rematch/immer"
-import models from "./models"
-
-// import type { RematchStore } from "@rematch/core"
-// @todo any used to be RematchStore<typeof models> but this isn't exactly right..
-// need to revisit this later. newer versions of rematch sorted TS out but
-// there are a lot of breaking changes for us it seems
-const store: any = init({
-  models,
-  plugins: [immerPlugin()],
-})
+import models, { IRootModel } from "./models"
+
+import { AsyncLocalStorage } from "async_hooks"
+import { IPluginOptions } from "./models/gatsby-api"
+import { GatsbyNodeApiHelpers } from "./utils/gatsby-types"
+
+export interface IGatsbyApiHook {
+  (helpers: GatsbyNodeApiHelpers, pluginOptions: IPluginOptions): Promise<void>
+}
+
+export type Store = RematchStore<IRootModel, Record<string, never>>
+
+export const asyncLocalStorage = new AsyncLocalStorage<Store>()
+
+const STORE_MAP = new Map<string, Store>()
+
+/**
+ * Wraps the API hook with the async local storage context
+ */
+
+export const wrapApiHook =
+  (hook: IGatsbyApiHook): IGatsbyApiHook =>
+  async (helpers, pluginOptions) => {
+    const typePrefix = pluginOptions.schema?.typePrefix ?? ``
+    if (!STORE_MAP.has(typePrefix)) {
+      STORE_MAP.set(
+        typePrefix,
+        init({
+          models,
+          plugins: [immerPlugin<IRootModel>()],
+        })
+      )
+    }
+
+    const store = STORE_MAP.get(typePrefix)
+
+    return asyncLocalStorage.run(store, async () =>
+      hook(helpers, pluginOptions)
+    )
+  }
+
+const store = (): Store => asyncLocalStorage.getStore()
```

Comment:
We change `store` from an exported store to a function which returns the correct store for the context